lib/matplotlib/axis.py

```diff
@@ -1028,62 +1028,29 @@ def _update_ticks(self, renderer):
-            # exceptions might arise from a user's projection (their
-            # rendition of the Axis object).  So, we catch all, with
-            # the idea that one would rather potentially lose a tick
-            # from one side of the axis or another, rather than see a
-            # stack trace.
-            # We also catch users warnings here. These are the result of
-            # invalid numpy calculations that may be the result of out of
-            # bounds on axis with finite allowed intervals such as geo
-            # projections i.e. Mollweide.
-            with np.errstate(invalid='ignore'):
-                try:
-                    ds1 = self._get_pixel_distance_along_axis(
-                        interval_expanded[0], -0.5)
-                except Exception:
-                    cbook._warn_external("Unable to find pixel distance "
-                                         "along axis for interval padding of "
-                                         "ticks; assuming no interval "
-                                         "padding needed.")
-                    ds1 = 0.0
-                if np.isnan(ds1):
-                    ds1 = 0.0
-                try:
-                    ds2 = self._get_pixel_distance_along_axis(
-                        interval_expanded[1], +0.5)
-                except Exception:
-                    cbook._warn_external("Unable to find pixel distance "
-                                         "along axis for interval padding of "
-                                         "ticks; assuming no interval "
-                                         "padding needed.")
-                    ds2 = 0.0
-                if np.isnan(ds2):
-                    ds2 = 0.0
-            interval_expanded = (interval_expanded[0] - ds1,
-                                 interval_expanded[1] + ds2)
+        if interval[1] <= interval[0]:
+             interval = interval[1], interval[0]
+        inter = self.get_transform().transform(interval)
 
         ticks_to_draw = []
         for tick, loc, label in tick_tups:
+            # draw each tick if it is in interval.   Note the transform
+            # to pixel space to take care of log transforms etc.
+            # interval_contains has a floating point tolerance.
             if tick is None:
                 continue
             # NB: always update labels and position to avoid issues like #9397
             tick.update_position(loc)
             tick.set_label1(label)
             tick.set_label2(label)
-            if not mtransforms.interval_contains(interval_expanded, loc):
+            try:
+                loct = self.get_transform().transform(loc)
+            except AssertionError:
+                # transforms.transform doesn't allow masked values but
+                # some scales might make them, so we need this try/except.
+                loct = None
+                continue
+            if not mtransforms.interval_contains_close(inter, loct):
```

Comment:
Do you have examples where we need `interval_contains_close` and not `interval_contains` here?  Why couldn't the *locators* be in charge of making sure whatever they return is in the interval and adjusting them if needed (except for those outside the interval needed for classic mode)?

I don't really like the idea of hardcoding the 1e-10 tolerance here when it's something that can clearly(?) be avoided.

Reply:
Hah, we currently have a 1/2-pixel tolerance!  Surely `1e-10 * interval` is preferable?   

The problem with limits is when people compute them.  

On my machine:

```
3.51-1.51 == 2.0
False

In [17]: abs(3.51-1.51- 2.0) < 1e-10
Out[17]: True
```

Reply:
I guess it doesn't make things worse, but at least let's keep interval_contains_close private?

The point is that the Locator API should say something like the following to implementers (including 3rd party locators):
> You may return tick values inside or outside of the axis bounds, but any value outside of the bounds (with no tolerance) will be dropped (except in "classic" mode, where we will keep the first outside value on either side).

and the locator classes are the one which should do something like
```
xmin, xmax = ...
ticks = ...
if xmin - tol < ticks[0] < xmin: ticks[0] = xmin
```
(or ticks[1] if that's the one corresponding to xmin).
i.e. the locator displaces the tick to match xmin if it's just below xmin.

Reply:
The issue with locators is that some uses want the bracketing ticks, unfortunately, for contour for instance...  We could fix/change that as well, but beyond this PR.

Reply:
Agreed it's beyond this PR.
The design above does not preclude bracketing ticks.

Reply:
@anntzer did you really want `interval_contains_close` to be private?  I actually think its useful to take into account floating point slop, but don't feel strongly about it.

Reply:
yes, it feels like a hack while waiting for locators to be fixed, but I'm not going to insist on it.

Reply:
Okey doke - done!